File: leetcode/367-valid-perfect-square/main.go
package main

import (
	"fmt"
)

// binary search between 1 and num
func isPerfectSquare(num int) bool {
	if num <= 0 {
		return false
	}
	min := 1
	max := num
	for min <= max {
		mean := (min + max) / 2
		if mean*mean == num {
			return true
		} else if mean*mean > num {
			max = mean - 1
		} else if mean*mean < num {
			min = mean + 1
		}
	}
	return false
}

func main() {
	fmt.Println(isPerfectSquare(1))
}

File: leetcode/367-valid-perfect-square/main.py
"""
    1st approach: binary search betweeen 1 and num, 1 <= x <= num

    Time    O(logn)
    Space   O(1)
    16 ms, faster than 92.52%
"""


class Solution(object):
    def isPerfectSquare(self, num):
        """
        :type num: int
        :rtype: bool
        """
        left = 1
        right = num
        while left <= right:
            mid = (left + right)/2
            if mid*mid < num:
                left = mid + 1
            elif mid*mid > num:
                right = mid - 1
            else:
                return True
        return False
